### PR TITLE
touch: change the error message to match the GNU error message #2346

### PR DESCRIPTION
**Pull request by Foryah**
I'm in the process of adding a test as well but I'm having problems running it. I've asked for help in the discord channel so as soon as I make it work, I'll also commit it.

### CODE REVIEW

File: src/uu/touch/src/touch.rs [hunk 1]
```diff
@@ -165,8 +165,8 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 continue;
             }
 
-            if let Err(e) = File::create(path) {
-                show_warning!("cannot touch '{}': {}", path, e);
+            if let Err(_) = File::create(path) {
+                show_error!("cannot touch '{}': {}", path, "No such file or directory");
```
Comment (miDeb) [inline src/uu/touch/src/touch.rs:169]: Can this really only happen due to a "No such file or directory error"? For example, I think "permission denied" could also cause the creation to fail.

**Comment by miDeb** [comment]
A test could be:
```rust
#[test]
fn test_error_msg() {
    new_ucmd!()
        .arg("nonexistent_dir/file")
        .fails()
        .stderr_only("touch: cannot touch 'nonexistent_dir/file': No such file or directory");
}
```

**Comment by Foryah** [comment]
Altered the solution to take into consideration permission denied errors and fallback on the error with ` (os error X)` and added tests for both cases.

**Comment by tertsdiepraam** [comment]
Tests are failing on Windows, probably because you used a unix-style path. Maybe try building the path with `std::path::PathBuf`? Another option is to create a file in the test and set the permission directly and test on that file. Something like this (but disabling read permission instead of write permission):
```rust
// create dest without write permissions
let mut permissions = at
    .make_file(TEST_HELLO_WORLD_DEST)
    .metadata()
    .unwrap()
    .permissions();
permissions.set_readonly(true);
set_permissions(at.plus(TEST_HELLO_WORLD_DEST), permissions).unwrap();
```

**Comment by jhscheer** [comment]
> However, Mac is also failing and I'm not sure why. Maybe it's because you're also using a non-existent file and Mac prioritizes that error over permission denied?

`/root/` does not exist on macOS.

> Another option is to create a file in the test and set the permission directly and test on that file.

This is the way :)

**Comment by Foryah** [comment]
Interesting discovery (at least for me): I created a file, gave if `0444` then tried to `touch` it and expected it to fail. It doesn't... It works like a charm. Then I thought "Of course it works, cause touch just opens the file and the OS actually changes the access time. So as long as you have read access, touch will work". Then I checked the options `touch` has and noticed a `--date` that allows you to set a specific access time. Then I thought "Ok, so in this case, touch manually changes the access time instead of relying on the OS. So for sure this will fail for a read only file". It doesn't... So then I went ahead and `chmod 0000` the file and tried `touch` again. Works just fine. So now I'm confused about how touch works 😁

I know how I can test the permission denied thought. I need to create a directory and restrict permissions on the directory not the file. Then touch will not be able to reach the file and throw a "Permission denied" error.

**Comment by jhscheer** [comment]
> Interesting discovery (at least for me): I created a file, gave if `0444` then tried to `touch` it and expected it to fail. It doesn't... It works like a charm. Then I thought "Of course it works, cause touch just opens the file and the OS actually changes the access time. So as long as you have read access, touch will work". Then I checked the options `touch` has and noticed a `--date` that allows you to set a specific access time. Then I thought "Ok, so in this case, touch manually changes the access time instead of relying on the OS. So for sure this will fail for a read only file". It doesn't... So then I went ahead and `chmod 0000` the file and tried `touch` again. Works just fine. So now I'm confused about how touch works grin

There's a difference between meta data and the file itself. [See also](https://stackoverflow.com/a/3385248)

**Comment by miDeb** [comment]
I am by no way an expert when it comes to file system permissions, but I think if you own the file you can do anything with it... Otherwise you could create a file that can no longer be deleted if you change permissions.
According to [GNU's codumentation](https://www.gnu.org/software/coreutils/manual/html_node/touch-invocation.html):
> When setting file timestamps to the current time, touch can change the timestamps for files that the user does not own but has write permission for. Otherwise, the user must own the files.

**Comment by Foryah** [comment]
Seems like the windows tests are still failing :( Guess folder permission is not enough to trigger the "Permission denied" error in windows...

**Comment by tertsdiepraam** [comment]
Alright, then I guess it's fine to put `#[cfg(unix)]` on this test, so we skip it on Windows.

File: tests/by-util/test_touch.rs [hunk 2]
```diff
@@ -466,3 +467,37 @@ fn test_touch_trailing_slash() {
     let file = "no-file/";
     ucmd.args(&[file]).fails();
 }
+
+#[test]
+fn test_touch_no_such_file_error_msg() {
+    let dirname = "notexistent";
```
Comment (tertsdiepraam) [inline tests/by-util/test_touch.rs:473]: `notexistent` --> `nonexistent` (to please the spell checker)

File: tests/by-util/test_touch.rs [hunk 3]
```diff
@@ -466,3 +467,37 @@ fn test_touch_trailing_slash() {
...
+        path_str
+    ));
+}
+
+#[test]
+#[cfg(unix)]
+fn test_touch_permission_denied_error_msg() {
+    let (at, mut ucmd) = at_and_ucmd!();
+
+    let dirname = "dir_with_read_only_acccess";
```
Comment (tertsdiepraam) [inline tests/by-util/test_touch.rs:489]: `dir_with_read_only_acccess` -> `dir_with_read_only_access`

**Review by tertsdiepraam — COMMENTED** [review]
Nice! All that's left are some spell checker warnings :)

**Comment by tertsdiepraam** [comment]
Thanks for your contribution!

**Comment by Foryah** [comment]
Glad I could contribute. I feel like you guys did most of the work anyways :) Thanks for all the help!

**Pull request merged**